relayer/workers/ethrelayer/main.go

```diff
@@ -0,0 +1,165 @@
+package ethrelayer
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/sirupsen/logrus"
+	"github.com/snowfork/go-substrate-rpc-client/v2/types"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/ethereum"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/parachain"
+	"github.com/snowfork/polkadot-ethereum/relayer/crypto/sr25519"
+)
+
+type Worker struct {
+	parachainConfig  *parachain.Config
+	ethereumConfig   *ethereum.Config
+	parachainConn    *parachain.Connection
+	ethereumConn     *ethereum.Connection
+	ethereumListener *EthereumListener
+	parachainWriter  *ParachainWriter
+	log              *logrus.Entry
+}
+
+const Name = "eth-relayer"
+
+func NewWorker(ethereumConfig *ethereum.Config, parachainConfig *parachain.Config) (*Worker, error) {
+	log := logrus.WithField("worker", Name)
+
+	log.Info("Creating worker")
+
+	ethereumConn := ethereum.NewConnection(ethereumConfig.Endpoint, nil, log)
+
+	// Generate keypair from secret
+	parachainKeypair, err := sr25519.NewKeypairFromSeed(parachainConfig.PrivateKey, "")
+	if err != nil {
+		return nil, err
+	}
+	parachainConn := parachain.NewConnection(parachainConfig.Endpoint, parachainKeypair.AsKeyringPair(), log)
+
+	// channel for messages from ethereum
+	ethMessages := make(chan []chain.Message, 1)
+	// channel for headers from ethereum (it's a blocking channel so that we
+	// can guarantee that a header is forwarded before we send dependent messages)
+	ethHeaders := make(chan chain.Header)
+
+	ethereumListener, err := NewEthereumListener(ethereumConfig, ethereumConn, ethMessages, ethHeaders, log)
+	if err != nil {
+		return nil, err
+	}
+
+	parachainWriter, err := NewParachainWriter(parachainConn, ethMessages, ethHeaders, log)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Worker{
+		ethereumConfig:   ethereumConfig,
+		ethereumConn:     ethereumConn,
+		parachainConfig:  parachainConfig,
+		parachainConn:    parachainConn,
+		ethereumListener: ethereumListener,
+		parachainWriter:  parachainWriter,
+		log:              log,
+	}, nil
+
+}
+
+func (worker *Worker) Start(ctx context.Context, eg *errgroup.Group) error {
+	worker.log.Info("Starting worker")
+
+	if worker.ethereumListener == nil || worker.parachainWriter == nil {
+		return fmt.Errorf("Sender and/or receiver need to be set before starting chain")
+	}
+
+	err := worker.ethereumConn.Connect(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Short-lived channels that communicate initialization parameters
+	// between the two chains. The chains close them after startup.
+	subInit := make(chan chain.Init)
+	ethInit := make(chan chain.Init)
+
+	// If the Substrate chain needs init params from Ethereum,
+	// retrieve them here and send to subInit before closing.
+	close(subInit)
+
+	eg.Go(func() error {
```

Comment:
You can get rid of this goroutine and simply use the value returned from `queryEthereumInitParams()` directly. See implementation in https://github.com/Snowfork/polkadot-ethereum/pull/354. Feel free to leave it to me as well. I understand if you don't want to change logic in these migrations

Reply:
yeh, will leave it to you, thanks!